tree.go

```diff
@@ -575,75 +604,115 @@ func (n *InternalNode) clearCache() {
 }
 
 func (n *LeafNode) Insert(k []byte, value []byte) error {
-	n.key = k
-	n.value = value
+	// Sanity check: ensure the key header is the same:
+	if bytes.Compare(k[:31], n.key[:31]) != 0 {
+		return errors.New("split should not happen here")
+	}
+	n.values[k[31]] = value
+	n.commitment = nil
 	return nil
 }
 
 func (n *LeafNode) InsertOrdered(key []byte, value []byte, flush chan FlushableNode) error {
-	err := n.Insert(key, value)
-	if err != nil && flush != nil {
-		flush <- FlushableNode{n.Hash(), n}
-	}
-	return err
+	// In the previous version, this value used to be flushed on insert.
+	// This is no longer the case, as all values at the last level get
+	// flushed at the same time.
+	return n.Insert(key, value)
 }
 
 func (n *LeafNode) Delete(k []byte) error {
-	return errors.New("cant delete a leaf in-place")
+	// Sanity check: ensure the key header is the same:
+	if bytes.Compare(k[:31], n.key[:31]) != 0 {
+		return errors.New("trying to delete a non-existing key")
+	}
+
+	n.values[k[31]] = nil
+	return nil
 }
 
 func (n *LeafNode) Get(k []byte, _ NodeResolverFn) ([]byte, error) {
-	if !bytes.Equal(k, n.key) {
+	if !bytes.Equal(k[:31], n.key[:31]) {
 		// If keys differ, return nil in order to
 		// signal that the key isn't present in the
 		// tree. Do not return an error, thus matching
 		// the behavior of Geth's SecureTrie.
 		return nil, nil
 	}
-	return n.value, nil
+	// value can be nil, as expected by geth
+	return n.values[k[31]], nil
 }
 
 func (n *LeafNode) ComputeCommitment() *bls.G1Point {
-	panic("can't compute the commitment directly")
+	if n.commitment != nil {
+		return n.commitment
+	}
+
+	emptyChildren := 0
```

Comment:
The code below can be turned into a function to be called by both internal and leaf nodes